File: py/cccount.py
#!/usr/bin/env python2
import sys
import os
import codecs
from optparse import OptionParser

class UI:
    def __init__ (self):
        #option settings
        op=OptionParser(usage='usage: %prog [OPTION]... [FILE]...', version="%prog 0.0.1")
        op.description="Print counts of Chinese characters in given FILE(s). If no"+\
                    " FILE is given,it will read from standard input stream"
        op.add_option('-l','--list',dest='list', 
                help = 'List all appeared Chinese characters with their counts',
                action = 'store_true')
        op.add_option('-s','--sum',dest='sum',
                help = 'Print the sum of all used Chinese characters',
                action = 'store_true')
        op.add_option('-e','--encoding',dest='encoding',default='utf8',
                help = 'Specify input encoding (default UTF-8)',
                action = 'store',type='string')
        self.ops,self.args = op.parse_args()
        
        #the default job is to print the sum 
        if self.ops.list is None and self.ops.sum is None:
            self.ops.sum = True
        
        #check up encoding
        try:
            codecs.lookup(self.ops.encoding)
        except LookupError as e:
            print( e)
            exit(1)
    
    def process(self):
        self.counter = Counter(encoding = self.ops.encoding)
        if len(self.args)==0:
            #read text from stdin
            self.counter.addFile(sys.stdin)
        else:
            #read text from file(s), whose name(s) are provided by args
            for f in self.args:
                if os.path.isfile(f):
                    with open(f) as fi:
                        self.counter.addFile(fi)
            
        self.__showResult()
        
    def __showResult(self):
        if self.ops.list:
           self.counter.lsChars()        
        if self.ops.sum:
           self.counter.showSum()
        

            
    
            
class Counter(object):
    def __init__(self,encoding='utf8'):
        self.cjks={}
        self.subject=None
        self.encoding=encoding
        self.done=False
        self.nfile=0 #number of files already processed
        self.ncjk=0 #number of Chinese characers appeared
        self.nchar=0 #number of chars already read

    
    def count(self):
        """do a counting on subject"""
        if not self.done:
            self.nfile += 1
            for l in self.subject:
                try:
                    line = unicode(l,self.encoding)
                except UnicodeDecodeError, e:
                    if hasattr(self.subject,"name"):
                        print '[%s]:' % self.subject.name
                    print  e
                    exit(1)
                for c in line:
                    self.nchar += 1
                    if (u'\u4e00' <= c and c <= u'\u9fa5') or\
                        (u'\u3400'<= c and c <= u'\u4db5') or\
                        (u'\U00020000' <= c and c <= u'\U0002a6d6'):
                        #Chars out of BMP use u'Uxxxxxxxx' notation
                        self.ncjk += 1
                        self.cjks[c]=self.cjks.get(c,0) + 1
        self.done = True            
        
    def lsChars(self):
        for c, n in sorted(self.cjks.iteritems(), key = lambda it:it[1],reverse = True):
            print c.encode('utf8'),n
   
    def addFile(self,f):
        """add one more file as the subject to be counted"""        
        self.done = False
        self.subject = f
        self.count()
        return self
         
        
    
    def showSum(self):
        """sum"""
        print "FILES: %d, CHINESE_CHARACTERS: %d, UNIQUE_APPEARANCES: %d " %\
                (self.nfile,self.ncjk,len(self.cjks))

if __name__=="__main__":
    UI().process()

File: ruby/cccount.rb
#! /usr/bin/env ruby
# encoding: utf-8
class CharIterator

  include Enumerable

  #@param inStream [#each_line]
  #@param inEncoding [Encoding]
  def initialize(inStream, inEncoding = Encoding::UTF_8)
    @ins = inStream
    @conv = inEncoding == Encoding::UTF_8 ? lambda{|s|s} : lambda{
        conv = Encoding::Converter.new( inEncoding, Encoding::UTF_8)
        conv.method :convert
    }[]
  end

  #@yieldparam [Fixnum]
  def each
    @ins.each_line do |line|
      line = @conv[line]
      line.each_codepoint{|cp| yield cp}
    end
  end
end



class Counter

  Cjk0 = 0x4e00..0x9fcb
  CjkA = 0x3400..0x4db5
  CjkB = 0x20000..0x2a6d6
  CjkC = 0x2a700..0x2b734
  CjkD = 0x2b740..0x2b81d

  #@param inEncoding [Encoding]
  def initialize(inEncoding=Encoding::UTF_8)
    @nChr, @nFiles, @nCjks, @cjk, @others = 0,0,0,0.upto(4).map{Hash.new(0)},Hash.new(0)
    @enc = inEncoding
  end

  attr_reader :nChr,:nFiles,:others

  def nCjks
    @cjk.map{|part|part.values.reduce(0,:+)}.reduce(0,:+)
  end

  def process(inStream)
    it = CharIterator.new inStream, @enc
    @nFiles += 1
    it.each { |c|
      @nChr += 1
      if c > 127
        case c
          when Cjk0
            @cjk[0][c] += 1
          when CjkA
            @cjk[1][c] += 1
          when CjkB
            @cjk[2][c] += 1
          when CjkC
            @cjk[3][c] += 1
          when CjkD
            @cjk[4][c] += 1
	        else
		        @others[c] += 1
        end
      else
	      @others[c] += 1
      end
    }
	  self
  end

	def sortedList
		@sortedList ||= @cjk.map{|h|h.sort_by{|ent|[-ent[1],ent[0]]}}.reduce(:concat)
	end

  #@param outStream [#<<]
	def listStr
		sortedList.map{|ent|"%c\t%d" % ent}.join("\n")
	end
  def othersStr
	  @others.sort_by{|ent|[-ent[1],ent[0]]}.map{|ent| "%c\t%d" % ent}.join("\n")
  end
  def nDistinguishable
     @cjk.reduce(0){|a,b|a + b.length}
  end

  def Cjks
	  @cjk.map(&:keys).reduce(:+)
  end

  def summary
    "#[Summary]:
    %8d -- Total number of characters have been read.
    %8d -- Total number of Chinese characters.
    %8d -- Number of distinguishable Chinese characters.
    %8d -- Files processed.
    " % [nChr,nCjks, nDistinguishable,nFiles]
  end
end

class CCCountOptParser
	def self.parse (args)
		scriptName = File.basename(__FILE__)
		options = OpenStruct.new
		options.fromEncoding = Encoding::UTF_8
		options.toEncoding = Encoding::UTF_8
		options.summary = true
		options.list = false
		options.others = false
		options.difference = false
		options.intersect = false

		opts = OptionParser.new do |opts|
			opts.banner = "Usage: #{scriptName} [OPTIONS] [FILE[S]]"
			opts.separator ''
			opts.separator "Specific options:"
			opts.on("-f","--from ENCODING","Specify the encoding of input text, default is utf-8"
			){|enc|
		    options.fromEncoding = Encoding::find enc
			}
			opts.on("-t","--to ENCODING","Specify the encoding of output text, defaut is utf-8"
			){|enc|
				options.toEncoding = Encoding::find enc
			}
			opts.on("-l","--[no-]list","List all appeared distinguishable Chinese characters," +
						" with corresponding counting, sorted by the counts descendantly."
			){|l|
				options.list = l
			}
			opts.on("-s","--[no-]summary","Show counting summary."
			){|s| options.summary = s}
			opts.on("-o","--[no-]others","List other characters besides cjk"
			){|o| options.others = o}
			opts.on("-d","--[no-]difference","Make a difference between cjk charsets of 2 files"
			){|d| options.difference = d}
			opts.on("-i","--[no-]intersect","Make a intersection between cjk charsets of 2 files"
			){|i| options.intersect = i}

		end
		opts.parse!(args)
		options
	end
end
class CCCountCLI

	# @param args [Array]
	def initialize (args = nil)
		args ||= ARGV.clone
		@opts = CCCountOptParser.parse(args)
		@files = args
		if @opts.difference || @opts.intersect
			if @files.length != 2
				puts "You have give 2 files to do difference or intersection"
				exit
			end
			fs=[]
			0.upto(1){|i|
				fs << Counter.new(@opts.fromEncoding).process(File.new(@files[i]))
			}
			lst = @opts.difference ? 	fs[0].Cjks - fs[1].Cjks : fs[0].Cjks & fs[1].Cjks
			puts lst.map{|c|c.chr @opts.toEncoding}.join('\n')

		else
			counter = Counter.new(@opts.fromEncoding)
			if args.length == 0
				counter.process $stdin
			else
				@files.each { |fname|
					File.open fname do |file|
						counter.process file
					end
				}
			end



			#show result
			if @opts.list
				puts counter.listStr.encode(@opts.toEncoding)
			end
			if @opts.others
				puts counter.othersStr.encode(@opts.toEncoding)
			end
			if @opts.summary
				puts counter.summary.encode(@opts.toEncoding)
			end
		end
	end

end

if $0 == __FILE__

	require "ostruct"
	require "optparse"
	CCCountCLI.new(ARGV)

end